src/spec.c

```diff
@@ -81,6 +81,59 @@ static void setMemoryInfo(RedisModuleCtx *ctx) {
   RedisModule_FreeServerInfo(ctx, info);
 }
 
+#define MILLISECOND_IN_ONE_SECOND 1000
+#define NANOSECOND_IN_ONE_MILLISECOND 1000000
+
+static inline timespec timespecFromMilliseconds(int64_t totalMilliseconds) {
+  timespec result = {.tv_sec = 0, .tv_nsec = 0};
+  if (totalMilliseconds > 0) {
+    result.tv_sec = totalMilliseconds / MILLISECOND_IN_ONE_SECOND;
+    result.tv_nsec = (totalMilliseconds % MILLISECOND_IN_ONE_SECOND) * NANOSECOND_IN_ONE_MILLISECOND;
+  }
+  return result;
+}
+
+static inline t_expirationTimePoint getDocExpirationTime(RedisModuleCtx* ctx, RedisModuleString *keystr) {
+  t_expirationTimePoint zero = {.tv_sec = 0, .tv_nsec = 0};
+  RedisModuleKey *openedKey = RedisModule_OpenKey(ctx, keystr, REDISMODULE_READ);
+  if (!openedKey) {
+    return zero;
```

Comment:
If we can't get here, or shouldn't, consider asserting instead of introducing uncovered code that will probably never be covered unless there was a fatal bug.
If you don't think we should crash here at least add a warning log.
I think we should crash here.

Reply:
If the expiration time is very small, e.g 1 millisecond.
And it was lazily expired, i.e we didn't get a notification for it.
Then it is possible that the opened key will be null, meaning it was expired...
It is an edge case but it is possible.

Reply:
Not sure I follow.
Can you try to make it happen? Let's understand if this is feasible before introducing it since we may just be getting unexpected behavior for that case.

Reply:
if it takes 1 millisecond for us to handle the notification then by the time we try an open the key it may already be lazy expired...
It is a race condition with the clock.
I can cause it by adding a sleep in the notification handling but not sure we want to do that.

Reply:
Let's discuss this.

Reply:
Fix discussed was to have this done when we already have the key open when loading the key data to our local struct (or before in the case of the hash fields).